File: scripts/coverage-report.py
#!/usr/bin/env python3
"""Filter and summarise the lcov report grcov produces for `cov`.

Run by `cov` (see `devenv.nix`) on grcov's raw output. Three jobs, one parse:

**Strip `#[cfg(test)]` modules.** Source-based coverage instruments the test
modules along with the code they test, and in this tree that is 39% of all
Rust lines. Left in, they do not merely inflate the headline — they invert the
signal that matters. `.claude/testing.md` requires every change to ship with a
test, so *every* PR here is production-code-plus-test-module, and a PR adding
100 untested production lines beside 300 test lines scores ~75% patch coverage
and reads as well tested. Neither Codecov's `ignore:` nor grcov's `--ignore`
can reach these lines: both match on paths, and these live inside production
source files.

Why not `#[coverage(off)]`, which is what the ecosystem points at: it is still
unstable on this repo's toolchain (rustc 1.96 gives `error[E0658]: the
`#[coverage]` attribute is an experimental feature`; tracking issue #84605, open
since 2021). Adopting it would mean a nightly toolchain for the coverage legs,
`#![feature(coverage_attribute)]` in 42 crate roots, and an annotation on all
184 `#[cfg(test)] mod tests` blocks that a new module can silently forget —
which is the same failure this exists to close, moved somewhere with 184 chances
to happen. When it stabilises the attribute becomes the better answer, because
the regions are then excluded at instrumentation time rather than after the
fact, and this file can go.

The exclusion is computed from the source rather than by regex-matching line
markers, because the obvious grcov spelling (`--excl-start '^#\\[cfg\\(test\\)\\]$'`
with `--excl-stop '^\\}$'`) is wrong on this tree in a direction that fails
silently. 14 of the 198 column-0 `#[cfg(test)]` sites are not module blocks —
`#[cfg(test)] mod foo;` declarations, a `use`, two `const`s — and for those the
range would run to whatever the next column-0 `}` happened to be, quietly
*deleting production lines* from the denominator. Requiring the `mod … {` form
and finding its own closing brace cannot do that.

**Rewrite the lcov in place of the raw one**, keeping every record it did not
have to drop and recomputing the `LF`/`LH`/`FNF`/`FNH`/`BRF`/`BRH` counters, so
Codecov reads exactly what the table below reports.

**Report**, and refuse to publish a report that measured nothing. A worst-first
per-crate table on stdout, `summary.json` with sorted keys and no timestamps for
agents and for diffing two runs, and the floors — nothing downstream of `cov`
can fail, so the floors and the named canary are the last thing standing between
a broken collection and a plausible number that reads as a drop.
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

CFG_TEST = "#[cfg(test)]"

# `mod <name> {` — the block form only, optionally `pub`/`pub(crate)`. A `mod
# foo;` declaration deliberately does not match: it opens no brace, so there is
# no range to find and guessing one deletes real code.
MOD_BLOCK = re.compile(r"^(?:pub(?:\([^)]*\))?\s+)?mod\s+[A-Za-z_][A-Za-z0-9_]*\s*\{\s*$")


def cfg_test_ranges(source: str) -> list[tuple[int, int]]:
    """Inclusive 1-based line ranges of the `#[cfg(test)]` modules in `source`.

    Only column-0 `#[cfg(test)]` immediately followed by a column-0 `mod … {`,
    closed by the next column-0 `}`. Everything inside such a module is indented
    by rustfmt, so the first column-0 `}` after it is its own — this does not
    need to balance braces, and deliberately does not try.
    """
    lines = source.splitlines()
    ranges: list[tuple[int, int]] = []
    i = 0
    while i < len(lines):
        if lines[i].rstrip() == CFG_TEST and i + 1 < len(lines) and MOD_BLOCK.match(lines[i + 1]):
            end = i + 2
            while end < len(lines) and lines[end].rstrip() != "}":
                end += 1
            if end < len(lines):
                ranges.append((i + 1, end + 1))
                i = end
        i += 1
    return ranges


def excluded(ranges: list[tuple[int, int]], line: int) -> bool:
    return any(start <= line <= stop for start, stop in ranges)


class FileRecord:
    """One `SF:` … `end_of_record` block, filtered."""

    def __init__(self, path: str) -> None:
        self.path = path
        # line -> hits, accumulated: a file can appear in more than one record
        # (one per binary that linked it), and a line is covered if *any* of
        # them hit it. Overwriting would report whichever binary was parsed last.
        self.lines: dict[int, int] = {}
        self.kept: list[str] = []
        self.functions: dict[str, int] = {}
        self.branches: list[tuple[int, bool]] = []
        self.dropped = 0

    @property
    def hit(self) -> int:
        return sum(1 for n in self.lines.values() if n > 0)

    @property
    def total(self) -> int:
        return len(self.lines)


def parse(raw: str, source_root: Path, strip_cfg_test: bool) -> dict[str, FileRecord]:
    """Parse an lcov report, accumulating per file and dropping excluded lines."""
    records: dict[str, FileRecord] = {}
    ranges_cache: dict[str, list[tuple[int, int]]] = {}
    current: FileRecord | None = None
    ranges: list[tuple[int, int]] = []
    dropped_fns: set[str] = set()

    for raw_line in raw.splitlines():
        line = raw_line.rstrip("\r")
        if line.startswith("SF:"):
            path = line[3:].strip()
            current = records.setdefault(path, FileRecord(path))
            dropped_fns = set()
            if path not in ranges_cache:
                source = source_root / path
                try:
                    text = source.read_text(encoding="utf-8", errors="replace")
                except OSError:
                    # grcov ran with --ignore-not-existing, so a path that
                    # cannot be read here is not one it reported lines for.
                    text = ""
                ranges_cache[path] = cfg_test_ranges(text) if strip_cfg_test else []
            ranges = ranges_cache[path]
            continue

        if current is None:
            continue

        if line.startswith("DA:"):
            number, _, rest = line[3:].partition(",")
            try:
                lineno = int(number)
                hits = int(rest.split(",", 1)[0])
            except ValueError:
                continue
            if excluded(ranges, lineno):
                current.dropped += 1
                continue
            current.lines[lineno] = current.lines.get(lineno, 0) + hits
            current.kept.append(f"DA:{lineno},{current.lines[lineno]}")
        elif line.startswith("FN:"):
            number, _, name = line[3:].partition(",")
            try:
                lineno = int(number)
            except ValueError:
                continue
            if excluded(ranges, lineno):
                # `FNDA:` is keyed by name, not by line, so the name has to be
                # remembered or its execution count outlives the function.
                dropped_fns.add(name)
                continue
            current.functions.setdefault(name, 0)
            current.kept.append(line)
        elif line.startswith("FNDA:"):
            count, _, name = line[5:].partition(",")
            if name in dropped_fns:
                continue
            try:
                current.functions[name] = current.functions.get(name, 0) + int(count)
            except ValueError:
                continue
        elif line.startswith("BRDA:"):
            number, _, rest = line[5:].partition(",")
            try:
                lineno = int(number)
            except ValueError:
                continue
            if excluded(ranges, lineno):
                continue
            taken = rest.rsplit(",", 1)[-1]
            current.branches.append((lineno, taken not in ("-", "0")))
            current.kept.append(line)
        elif line.startswith("end_of_record"):
            current = None

    return records


def write_lcov(records: dict[str, FileRecord], out: Path) -> None:
    """Emit the filtered report, with every counter recomputed.

    Recomputed rather than carried over: a stale `LF:`/`LH:` next to a filtered
    `DA:` list is a report that disagrees with itself, and consumers differ on
    which half they believe.
    """
    chunks: list[str] = []
    for path in sorted(records):
        record = records[path]
        if not record.lines:
            continue
        body = [f"SF:{path}"]
        for name, count in sorted(record.functions.items()):
            body.append(f"FNDA:{count},{name}")
        body.extend(line for line in record.kept if line.startswith(("FN:", "BRDA:")))
        body.append(f"FNF:{len(record.functions)}")
        body.append(f"FNH:{sum(1 for n in record.functions.values() if n > 0)}")
        for lineno in sorted(record.lines):
            body.append(f"DA:{lineno},{record.lines[lineno]}")
        body.append(f"LF:{record.total}")
        body.append(f"LH:{record.hit}")
        if record.branches:
            body.append(f"BRF:{len(record.branches)}")
            body.append(f"BRH:{sum(1 for _, taken in record.branches if taken)}")
        body.append("end_of_record")
        chunks.append("\n".join(body))
    out.write_text("\n".join(chunks) + "\n", encoding="utf-8")


def unit_of(path: str) -> str:
    """The reporting unit a source path belongs to.

    Crates are the granularity a human acts on: "coverage dropped" is not
    actionable, "coverage dropped in crates/engine" is.
    """
    parts = [part for part in path.split("/") if part not in ("", ".")]
    if not parts:
        return "<unknown>"
    if path.startswith("/"):
        # grcov's `--ignore '/*'` should have removed these; name them rather
        # than rendering a blank row if one ever survives.
        return "<external>"
    if parts[0] == "crates" and len(parts) > 1:
        return f"crates/{parts[1]}"
    if parts[0] == "src":
        # The root `heph` package's own sources.
        return "heph"
    return parts[0]


def pct(hit: int, total: int) -> float:
    return round(100.0 * hit / total, 2) if total else 0.0


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("lcov", type=Path, help="grcov's raw lcov report")
    ap.add_argument(
        "--source-root",
        type=Path,
        default=Path("."),
        help="root the lcov paths are relative to",
    )
    ap.add_argument("--out-lcov", type=Path, help="write the filtered report here")
    ap.add_argument("--json", type=Path, help="write a summary.json here")
    ap.add_argument(
        "--strip-cfg-test",
        action="store_true",
        help="drop lines inside column-0 `#[cfg(test)] mod … {}` blocks",
    )
    ap.add_argument(
        "--label",
        default="",
        help="platform label printed with the totals (e.g. linux/amd64)",
    )
    # The three below are the floors that make an empty report fail rather than
    # publish. Nothing downstream of `cov` can: Codecov is configured
    # informational and the CI job gates nothing, so a report that measured
    # almost nothing would land as a plausible number and read as a drop.
    ap.add_argument(
        "--min-files",
        type=int,
        default=0,
        help="fail if fewer than this many source files appear in the report",
    )
    ap.add_argument(
        "--min-lines",
        type=int,
        default=0,
        help="fail if fewer than this many instrumented lines appear",
    )
    ap.add_argument(
        "--require-covered",
        action="append",
        default=[],
        metavar="PATH",
        help=(
            "fail unless this source path is present with at least one hit. A "
            "size floor can be met by build-script noise; a named file that "
            "must have been executed cannot."
        ),
    )
    args = ap.parse_args()

    raw = args.lcov.read_text(encoding="utf-8", errors="replace")
    records = parse(raw, args.source_root, args.strip_cfg_test)
    records = {path: record for path, record in records.items() if record.lines}

    if args.out_lcov:
        write_lcov(records, args.out_lcov)

    if not records:
        print(
            f"error: {args.lcov} names no source files with instrumented lines "
            f"— the report is empty, which is not the same thing as 0% coverage",
            file=sys.stderr,
        )
        return 1

    per_unit: dict[str, list[int]] = {}
    for path, record in records.items():
        acc = per_unit.setdefault(unit_of(path), [0, 0])
        acc[0] += record.hit
        acc[1] += record.total

    grand_hit = sum(record.hit for record in records.values())
    grand_total = sum(record.total for record in records.values())
    stripped = sum(record.dropped for record in records.values())

    # Worst first, then by name so equal percentages have a stable order (a
    # run-to-run reshuffle in a CI log reads as a change that isn't one).
    units = sorted(per_unit.items(), key=lambda kv: (pct(*kv[1]), kv[0]))

    for name, (hit, total) in units:
        print(f"{pct(hit, total):6.2f}%  {hit:7d}/{total:<7d}  {name}")
    label = f" [{args.label}]" if args.label else ""
    print(f"{pct(grand_hit, grand_total):6.2f}%  {grand_hit:7d}/{grand_total:<7d}  TOTAL{label}")
    if args.strip_cfg_test:
        print(f"({stripped} lines in #[cfg(test)] modules excluded)")

    if args.json:
        payload = {
            "label": args.label,
            "stripped_cfg_test_lines": stripped,
            "totals": {
                "line_coverage": pct(grand_hit, grand_total),
                "lines_hit": grand_hit,
                "lines_total": grand_total,
            },
            "units": [
                {
                    "line_coverage": pct(hit, total),
                    "lines_hit": hit,
                    "lines_total": total,
                    "name": name,
                }
                for name, (hit, total) in units
            ],
            "files": [
                {
                    "line_coverage": pct(record.hit, record.total),
                    "lines_hit": record.hit,
                    "lines_total": record.total,
                    "path": path,
                }
                for path, record in sorted(records.items())
            ],
        }
        args.json.write_text(
            json.dumps(payload, indent=2, sort_keys=True) + "\n", encoding="utf-8"
        )

    # Checked after the table is printed, so a failure still shows what *was*
    # measured — the first thing needed to tell "collection broke" from "a crate
    # stopped being built".
    problems = []
    if len(records) < args.min_files:
        problems.append(f"only {len(records)} source files in the report (floor: {args.min_files})")
    if grand_total < args.min_lines:
        problems.append(
            f"only {grand_total} instrumented lines in the report (floor: {args.min_lines})"
        )
    for required in args.require_covered:
        record = records.get(required)
        if record is None:
            problems.append(f"{required} is absent from the report entirely")
        elif record.hit == 0:
            problems.append(f"{required} has 0 of {record.total} lines covered")

    if problems:
        print(file=sys.stderr)
        print(
            "error: this report is too small to be a coverage measurement — "
            "collection failed, it is not a drop:",
            file=sys.stderr,
        )
        for problem in problems:
            print(f"  - {problem}", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/coverage_report.rs
//! Behavioural tests for `scripts/coverage-report.py`.
//!
//! `tests/coverage_gate.rs` asserts that `cov` *invokes* the floors. This file
//! asserts they work, which is a different claim and the one that matters:
//! nothing downstream of `cov` can fail — Codecov is informational and the job
//! gates nothing — so this script is the last thing standing between a broken
//! collection and a plausible number published as a drop. A gate that only
//! checks the flag is spelled correctly leaves that property resting on
//! nothing.
//!
//! Two behaviours are covered, both with failure modes that are silent by
//! construction:
//!
//!   - **the floors and the canary**, which must fire on an empty or tiny
//!     report and must *not* fire on a healthy one;
//!   - **`--strip-cfg-test`**, which deletes lines from the denominator. Too
//!     little and 39% of this tree's lines are test modules counted as covered
//!     production code; too much and it silently removes real code, which moves
//!     coverage in the flattering direction. Both directions are asserted.
//!
//! Fixtures are written into a `TempDir` rather than pointed at the real tree:
//! the parse has to be exercised against inputs that do not exist here yet
//! (a `DA:` with a checksum field, an unbalanced module, an absolute path).

use std::path::Path;
use std::process::{Command, Output};

fn repo_root() -> &'static Path {
    Path::new(env!("CARGO_MANIFEST_DIR"))
}

/// Run the script over `lcov`, with `source_root` as the tree its paths are
/// relative to.
fn run(dir: &Path, lcov: &str, args: &[&str]) -> Output {
    let lcov_path = dir.join("in.info");
    std::fs::write(&lcov_path, lcov).expect("write lcov fixture");
    Command::new("python3")
        .arg(repo_root().join("scripts/coverage-report.py"))
        .arg(&lcov_path)
        .arg("--source-root")
        .arg(dir)
        .args(args)
        .output()
        .expect("run coverage-report.py")
}

fn stdout(out: &Output) -> String {
    String::from_utf8_lossy(&out.stdout).into_owned()
}

fn stderr(out: &Output) -> String {
    String::from_utf8_lossy(&out.stderr).into_owned()
}

/// A minimal well-formed record for `path` with the given `line,hits` pairs.
fn record(path: &str, lines: &[(u32, u32)]) -> String {
    let mut out = format!("SF:{path}\n");
    for (line, hits) in lines {
        out.push_str(&format!("DA:{line},{hits}\n"));
    }
    out.push_str("end_of_record\n");
    out
}

#[test]
fn an_empty_report_fails_rather_than_reporting_zero_percent() {
    let dir = tempfile::tempdir().expect("tempdir");
    let out = run(dir.path(), "", &[]);

    assert!(
        !out.status.success(),
        "an lcov with no source files exited 0. That publishes 'collection \
         broke' as '0% coverage', and Codecov cannot tell the two apart."
    );
    assert!(
        stderr(&out).contains("not the same thing as 0% coverage"),
        "the empty-report failure does not say what happened:\n{}",
        stderr(&out)
    );
}

/// A file listed with only excluded lines must not survive as a phantom entry.
#[test]
fn a_record_left_with_no_lines_is_dropped_entirely() {
    let dir = tempfile::tempdir().expect("tempdir");
    std::fs::write(
        dir.path().join("only_tests.rs"),
        "#[cfg(test)]\nmod tests {\n    fn a() {}\n}\n",
    )
    .expect("write source");

    let out = run(
        dir.path(),
        &record("only_tests.rs", &[(2, 1), (3, 1)]),
        &["--strip-cfg-test"],
    );

    assert!(
        !out.status.success(),
        "a report whose every line was excluded exited 0 with nothing left to \
         measure:\n{}",
        stdout(&out)
    );
}

#[test]
fn the_floors_fire_and_say_which_one() {
    let dir = tempfile::tempdir().expect("tempdir");
    let lcov = record("src/a.rs", &[(1, 1), (2, 0)]);

    let out = run(
        dir.path(),
        &lcov,
        &["--min-files", "10", "--min-lines", "500"],
    );
    assert!(!out.status.success(), "the floors did not fire");
    let err = stderr(&out);
    assert!(
        err.contains("only 1 source files") && err.contains("only 2 instrumented lines"),
        "the floor failure does not name both breaches:\n{err}"
    );

    // The table still has to be printed, or a failure cannot be told apart
    // from a drop without re-running.
    assert!(
        stdout(&out).contains("TOTAL"),
        "the table was not printed on the failing path, so there is nothing to \
         tell 'collection broke' from 'a crate stopped being built':\n{}",
        stdout(&out)
    );
}

#[test]
fn the_floors_pass_at_the_boundary() {
    let dir = tempfile::tempdir().expect("tempdir");
    let out = run(
        dir.path(),
        &record("src/a.rs", &[(1, 1), (2, 0)]),
        &["--min-files", "1", "--min-lines", "2"],
    );
    assert!(
        out.status.success(),
        "a report exactly at the floors was rejected, so the floors are \
         off-by-one and will red a healthy run:\n{}",
        stderr(&out)
    );
}

/// The canary is the guard a size floor cannot be: a floor is satisfiable by
/// build-script noise, a named file that must have executed is not.
#[test]
fn require_covered_distinguishes_absent_from_unhit_from_covered() {
    let dir = tempfile::tempdir().expect("tempdir");
    let lcov = format!(
        "{}{}",
        record("src/hit.rs", &[(1, 3), (2, 1)]),
        record("src/cold.rs", &[(1, 0), (2, 0)])
    );

    let absent = run(dir.path(), &lcov, &["--require-covered", "src/gone.rs"]);
    assert!(!absent.status.success(), "an absent canary passed");
    assert!(
        stderr(&absent).contains("absent from the report entirely"),
        "{}",
        stderr(&absent)
    );

    let unhit = run(dir.path(), &lcov, &["--require-covered", "src/cold.rs"]);
    assert!(
        !unhit.status.success(),
        "a canary present with zero hits passed — which is exactly the shape of \
         a run where the binaries were built but never executed"
    );
    assert!(
        stderr(&unhit).contains("0 of 2 lines covered"),
        "{}",
        stderr(&unhit)
    );

    let covered = run(dir.path(), &lcov, &["--require-covered", "src/hit.rs"]);
    assert!(
        covered.status.success(),
        "a covered canary was rejected:\n{}",
        stderr(&covered)
    );
}

/// The same file appears once per binary that linked it. A line hit by only one
/// of those records is covered.
#[test]
fn duplicate_records_for_one_file_are_accumulated_not_overwritten() {
    let dir = tempfile::tempdir().expect("tempdir");
    let lcov = format!(
        "{}{}",
        record("src/a.rs", &[(1, 0), (2, 0)]),
        record("src/a.rs", &[(1, 4), (2, 0)])
    );

    let out = run(dir.path(), &lcov, &["--require-covered", "src/a.rs"]);
    assert!(
        out.status.success(),
        "a line hit only by the second record was reported as unhit, so \
         whichever binary parsed last decides the number:\n{}",
        stderr(&out)
    );
    assert!(
        stdout(&out).contains("      1/2"),
        "expected 1 of 2 lines covered after accumulation:\n{}",
        stdout(&out)
    );
}

/// lcov's `DA:` allows a third checksum field. Dropping those records would
/// shrink the denominator silently.
#[test]
fn da_records_with_a_checksum_field_are_parsed() {
    let dir = tempfile::tempdir().expect("tempdir");
    let lcov = "SF:src/a.rs\nDA:1,2,abc123\nDA:2,0,def456\nDA:oops,1\nend_of_record\n";

    let out = run(dir.path(), lcov, &["--min-lines", "2"]);
    assert!(
        out.status.success(),
        "checksum-bearing DA records were dropped, or the malformed one \
         crashed the parse:\n{}",
        stderr(&out)
    );
    assert!(
        stdout(&out).contains("      1/2"),
        "expected exactly the two well-formed lines:\n{}",
        stdout(&out)
    );
}

#[test]
fn strip_cfg_test_removes_the_test_module_and_keeps_production_code() {
    let dir = tempfile::tempdir().expect("tempdir");
    std::fs::write(
        dir.path().join("lib.rs"),
        // 1 pub fn real() -> u32 {
        // 2     1
        // 3 }
        // 4
        // 5 #[cfg(test)]
        // 6 mod tests {
        // 7     #[test]
        // 8     fn t() {}
        // 9 }
        "pub fn real() -> u32 {\n    1\n}\n\n#[cfg(test)]\nmod tests {\n    #[test]\n    fn t() {}\n}\n",
    )
    .expect("write source");

    let lcov = record("lib.rs", &[(1, 1), (2, 0), (7, 1), (8, 1)]);
    let out_lcov = dir.path().join("out.info");
    let out = run(
        dir.path(),
        &lcov,
        &[
            "--strip-cfg-test",
            "--out-lcov",
            out_lcov.to_str().expect("utf-8 path"),
        ],
    );
    assert!(out.status.success(), "{}", stderr(&out));

    let filtered = std::fs::read_to_string(&out_lcov).expect("read filtered lcov");
    assert!(
        filtered.contains("DA:1,1") && filtered.contains("DA:2,0"),
        "production lines were deleted from the denominator, which moves \
         coverage in the flattering direction:\n{filtered}"
    );
    assert!(
        !filtered.contains("DA:7,") && !filtered.contains("DA:8,"),
        "#[cfg(test)] lines survived; they are 39% of this tree and they invert \
         patch coverage:\n{filtered}"
    );
    assert!(
        filtered.contains("LF:2") && filtered.contains("LH:1"),
        "the LF/LH counters were not recomputed, so the report disagrees with \
         its own DA records:\n{filtered}"
    );
    assert!(
        stdout(&out).contains("2 lines in #[cfg(test)] modules excluded"),
        "the exclusion is invisible in the output:\n{}",
        stdout(&out)
    );
}

/// The reason this is not grcov's `--excl-start`/`--excl-stop`: a
/// `#[cfg(test)] mod foo;` declaration opens no brace, and a regex range would
/// run to the next column-0 `}` — deleting production code. 14 sites in this
/// tree are of that shape.
#[test]
fn a_cfg_test_declaration_or_item_excludes_nothing() {
    let dir = tempfile::tempdir().expect("tempdir");
    std::fs::write(
        dir.path().join("lib.rs"),
        // 1 #[cfg(test)]
        // 2 mod helpers;
        // 3
        // 4 pub fn real() -> u32 {
        // 5     1
        // 6 }
        "#[cfg(test)]\nmod helpers;\n\npub fn real() -> u32 {\n    1\n}\n",
    )
    .expect("write source");

    let out_lcov = dir.path().join("out.info");
    let out = run(
        dir.path(),
        &record("lib.rs", &[(4, 1), (5, 1)]),
        &[
            "--strip-cfg-test",
            "--out-lcov",
            out_lcov.to_str().expect("utf-8 path"),
        ],
    );
    assert!(out.status.success(), "{}", stderr(&out));

    let filtered = std::fs::read_to_string(&out_lcov).expect("read filtered lcov");
    assert!(
        filtered.contains("DA:4,1") && filtered.contains("DA:5,1"),
        "a `#[cfg(test)] mod foo;` declaration swallowed the code after it — \
         this is precisely the failure the regex approach has:\n{filtered}"
    );
}

/// An unterminated module must exclude nothing rather than run to EOF.
#[test]
fn an_unclosed_cfg_test_module_excludes_nothing() {
    let dir = tempfile::tempdir().expect("tempdir");
    std::fs::write(
        dir.path().join("lib.rs"),
        "pub fn real() {}\n#[cfg(test)]\nmod tests {\n    fn t() {}\n",
    )
    .expect("write source");

    let out_lcov = dir.path().join("out.info");
    let out = run(
        dir.path(),
        &record("lib.rs", &[(1, 1)]),
        &[
            "--strip-cfg-test",
            "--out-lcov",
            out_lcov.to_str().expect("utf-8 path"),
        ],
    );
    assert!(out.status.success(), "{}", stderr(&out));
    assert!(
        std::fs::read_to_string(&out_lcov)
            .expect("read filtered lcov")
            .contains("DA:1,1"),
        "an unclosed module excluded to EOF"
    );
}

/// Function records are keyed by line but their execution counts are keyed by
/// name, so dropping one without the other leaves a count for a function that
/// is no longer in the report.
#[test]
fn function_records_inside_a_test_module_are_dropped_with_their_counts() {
    let dir = tempfile::tempdir().expect("tempdir");
    std::fs::write(
        dir.path().join("lib.rs"),
        "pub fn real() {\n}\n#[cfg(test)]\nmod tests {\n    fn t() {}\n}\n",
    )
    .expect("write source");

    let lcov = "SF:lib.rs\nFN:1,real\nFN:5,tests::t\nFNDA:2,real\nFNDA:9,tests::t\n\
                DA:1,2\nDA:5,9\nend_of_record\n";
    let out_lcov = dir.path().join("out.info");
    let out = run(
        dir.path(),
        lcov,
        &[
            "--strip-cfg-test",
            "--out-lcov",
            out_lcov.to_str().expect("utf-8 path"),
        ],
    );
    assert!(out.status.success(), "{}", stderr(&out));

    let filtered = std::fs::read_to_string(&out_lcov).expect("read filtered lcov");
    assert!(
        !filtered.contains("tests::t"),
        "a test function's FNDA outlived its FN record:\n{filtered}"
    );
    assert!(
        filtered.contains("FNF:1") && filtered.contains("FNH:1"),
        "the function counters were not recomputed:\n{filtered}"
    );
}

/// Reporting units are what the table is actionable by; a blank one is a row
/// nobody can act on.
#[test]
fn reporting_units_are_named_for_every_path_shape() {
    let dir = tempfile::tempdir().expect("tempdir");
    let lcov = format!(
        "{}{}{}",
        record("crates/engine/src/a.rs", &[(1, 1)]),
        record("src/b.rs", &[(1, 1)]),
        record("/nix/store/x/c.rs", &[(1, 1)])
    );

    let out = run(dir.path(), &lcov, &[]);
    assert!(out.status.success(), "{}", stderr(&out));
    let table = stdout(&out);

    for unit in ["crates/engine", "heph", "<external>"] {
        assert!(
            table.contains(unit),
            "no `{unit}` row in the table — an absolute path that escapes \
             grcov's `--ignore '/*'` would render as a blank row:\n{table}"
        );
    }
}

/// `summary.json` is the agent surface: it has to be stable enough that two
/// runs of the same tree diff to nothing.
#[test]
fn summary_json_is_stable_and_carries_no_timestamp() {
    let dir = tempfile::tempdir().expect("tempdir");
    let lcov = record("src/a.rs", &[(1, 1), (2, 0)]);
    let json = dir.path().join("summary.json");

    let mut renders = Vec::new();
    for _ in 0..2 {
        let out = run(
            dir.path(),
            &lcov,
            &[
                "--json",
                json.to_str().expect("utf-8 path"),
                "--label",
                "linux/amd64",
            ],
        );
        assert!(out.status.success(), "{}", stderr(&out));
        renders.push(std::fs::read_to_string(&json).expect("read summary.json"));
    }

    assert_eq!(
        renders.first(),
        renders.last(),
        "summary.json is not reproducible across runs, so it cannot be diffed"
    );
    let rendered = renders.first().expect("one render");
    assert!(
        rendered.contains("\"line_coverage\": 50.0")
            && rendered.contains("\"label\": \"linux/amd64\""),
        "summary.json does not carry the totals it is read for:\n{rendered}"
    );
}
